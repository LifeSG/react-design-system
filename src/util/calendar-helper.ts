import dayjs, { Dayjs } from "dayjs";

export namespace CalendarHelper {
    // =============================================================================
    // HELPER FUNCTIONS
    // =============================================================================
    export const generateDays = (calendarDate: Dayjs): Dayjs[][] => {
        const firstDayOfTheMonth = calendarDate.startOf("month");

        const firstDayOfFirstWeekOfMonth =
            dayjs(firstDayOfTheMonth).startOf("week");

        const firstDayOfEachWeek = generateFirstDayOfEachWeek(
            firstDayOfFirstWeekOfMonth
        );

        return firstDayOfEachWeek.map((date) => generateWeek(date));
    };

    export const generateMonths = (calendarDate: Dayjs): Dayjs[] => {
        const months: Dayjs[] = [];

        for (let i = 0; i < 12; i++) {
            const monthForSelectedDay = calendarDate.month(i);
            months.push(dayjs(monthForSelectedDay));
        }

        return months;
    };

    export const generateDecadeOfYears = (calendarDate: Dayjs): Dayjs[] => {
        const year = calendarDate.year();
        const decade = Math.floor(year / 10) * 10;

        const base = calendarDate.year(decade);
        const prev = base.subtract(1, "year");

        const years = [prev, base];

        for (let i = 1; i < 11; i++) {
            years.push(base.add(i, "year"));
        }

        return years;
    };

    export const getStartEndDecade = (calendarDate: Dayjs) => {
        const beginDecade = Math.floor(+calendarDate.format("YYYY") / 10) * 10;
        const endDecade = +dayjs(`${beginDecade + 9}-01-01`).format("YYYY");

        return {
            beginDecade,
            endDecade,
        };
    };
}

const generateFirstDayOfEachWeek = (day: Dayjs): Dayjs[] => {
    const dates: Dayjs[] = [day];

    for (let i = 1; i < 6; i++) {
        const date = day.add(i, "week");
        dates.push(date);
    }

    return dates;
};

const generateWeek = (day: Dayjs): Dayjs[] => {
    const dates: Dayjs[] = [];

    for (let i = 0; i < 7; i++) {
        const date = day.add(i, "day");
        dates.push(date);
    }

    return dates;
};
