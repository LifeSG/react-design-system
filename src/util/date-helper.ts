export namespace DateHelper {
    export const clampDay = (
        day: string,
        month: string,
        year: string
    ): string => {
        const dayNumber = parseInt(day);
        const monthNumber = parseInt(month);
        const yearNumber = parseInt(year);

        if (dayNumber == 0) {
            return "1";
        }

        switch (monthNumber) {
            case 1:
            case 3:
            case 5:
            case 7:
            case 8:
            case 10:
            case 12: {
                if (dayNumber > 31) {
                    return "31";
                }
                break;
            }
            case 4:
            case 6:
            case 9:
            case 11:
                if (dayNumber > 30) {
                    return "30";
                }
                break;
            case 2:
                if (isLeapYear(yearNumber)) {
                    if (dayNumber > 29) {
                        return "29";
                    }
                } else {
                    if (dayNumber > 28) {
                        return "28";
                    }
                }
                break;
            default:
                break;
        }

        return day;
    };
    export const clampMonth = (month: string): string => {
        const monthNumber = parseInt(month);

        if (monthNumber == 0) {
            return "1";
        }

        if (monthNumber > 12) {
            return "12";
        }
        return month;
    };
    export const isLeapYear = (year: number): boolean => {
        return (year % 4 === 0 && year % 100 !== 0) || year % 400 === 0;
    };
}
