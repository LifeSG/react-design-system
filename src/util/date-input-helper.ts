import dayjs from "dayjs";
import isBetween from "dayjs/plugin/isBetween";

dayjs.extend(isBetween);

export namespace DateInputHelper {
    const dateFormat = /^\d{4}-\d{2}-\d{2}$/; // YYYY-MM-DD

    export const isDateDisabled = (
        val: string | undefined,
        props: { disabledDates?: string[]; between?: [string, string] }
    ): boolean => {
        const { disabledDates, between } = props;
        if (
            disabledDates &&
            disabledDates.length &&
            disabledDates.includes(val)
        ) {
            return true;
        }

        if (between && between.length === 2) {
            const [min, max] = between;
            if (!min || !max) {
                console.warn('"between" prop is invalid');
                return true;
            }

            if (!dayjs(val).isBetween(min, max, "day", "[]")) {
                return true;
            }
        }

        return false;
    };

    export const validate = (
        startDate: string | undefined,
        endDate: string | undefined,
        disabledDates?: string[] | undefined,
        between?: string[] | undefined
    ): boolean => {
        if (!startDate || !endDate) {
            return false;
        }

        if (dayjs(startDate).isAfter(endDate)) {
            return false;
        }

        if (
            disabledDates &&
            disabledDates.length &&
            disabledDates.some((value) => [startDate, endDate].includes(value))
        ) {
            return false;
        }

        if (between && between.length) {
            if (
                ![startDate, endDate].every((selectedDate) =>
                    dayjs(selectedDate).isBetween(
                        between[0],
                        between[1],
                        "day",
                        "[]"
                    )
                )
            ) {
                return false;
            }

            if (!between.every((value) => dateFormat.test(value))) {
                return false;
            }
        }

        return true;
    };

    export const validateSingle = (
        value: string,
        disabledDates?: string[] | undefined,
        between?: string[] | undefined
    ) => {
        if (value.length === 0) {
            return false;
        }

        if (
            disabledDates &&
            disabledDates.length &&
            disabledDates.some((disabledDate) => value === disabledDate)
        ) {
            return false;
        }

        if (
            between &&
            between.length &&
            !dayjs(value).isBetween(between[0], between[1], "day", "[]")
        ) {
            if (!dayjs(value).isBetween(between[0], between[1], "day", "[]")) {
                return false;
            }

            if (!between.every((value) => dateFormat.test(value))) {
                return false;
            }
        }

        return true;
    };

    export const getFormattedRawValue = (
        startValue: string,
        endValue?: string
    ): {
        start: string[];
        end: string[] | undefined;
    } => {
        const delimiter = "-";
        const [year, month, day] = startValue.split(delimiter);

        const start = [day, month, year];

        let end = undefined;
        if (endValue) {
            const [endYear, endMonth, endDay] = endValue.split(delimiter);
            end = [endDay, endMonth, endYear];
        }

        return {
            start,
            end,
        };
    };

    export const sleep = (ms: number, controller: AbortController) =>
        new Promise((resolve) => {
            const timeoutId = setTimeout(resolve, ms);

            controller.signal.addEventListener("abort", () => {
                clearTimeout(timeoutId);
            });
        });
}
