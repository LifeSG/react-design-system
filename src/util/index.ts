export * from "./calendar-helper";
export * from "./date-helper";
export * from "./date-input-helper";
export * from "./simple-id-generator";
export * from "./string-helper";
export * from "./use-container-query";
export * from "./use-event-listener";
export * from "./use-next-input-state";
export * from "./use-state-ref";
