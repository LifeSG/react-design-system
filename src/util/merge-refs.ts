import { MutableRefObject, Ref } from "react";
import { isReact19 } from "./version";

export const mergeRefs = <T>(...refs: Ref<T>[]) => {
    return (value: T | null) => {
        const cleanups: (() => void)[] = [];

        for (const ref of refs) {
            if (typeof ref === "function") {
                const cleanup = ref(value);
                const isCleanup = typeof cleanup === "function";
                cleanups.push(isCleanup ? cleanup : () => ref(null));
            } else {
                (ref as MutableRefObject<T | null>).current = value;
                cleanups.push(() => {
                    (ref as MutableRefObject<T | null>).current = null;
                });
            }
        }

        if (isReact19()) {
            return () => {
                for (const cleanup of cleanups) cleanup();
            };
        }
    };
};
