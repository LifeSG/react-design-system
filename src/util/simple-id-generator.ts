export namespace SimpleIdGenerator {
    /**
     * Generates a simple generator for element ids
     */
    export const generate = (): string => {
        return Math.random().toString(36).substring(2, 9);
    };
}
