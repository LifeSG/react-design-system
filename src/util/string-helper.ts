import { MaskAttributeProps } from "../masked-input";

export namespace StringHelper {
    export const transformWithSpaces = (
        value: string | any,
        groupLength: number
    ) => {
        const regex = "(.{" + groupLength + "})";
        return value
            .replace(/[^\dA-Z]/gi, "")
            .replace(new RegExp(regex, "g"), "$1 ")
            .trim();
    };

    /**
     * Pads single digit values with a zero and trims leading zero for
     * values of length more than 2
     * @param value the string value
     * @param retainSingleZero setting true will allow single zero value
     */
    export const padValue = (
        value: string,
        retainSingleZero?: boolean
    ): string => {
        if (value === "0") {
            return retainSingleZero ? value : value.padStart(2, "0");
        }

        if (value === "") {
            return value;
        }

        const trimmedValue = parseInt(value);
        return trimmedValue.toString().padStart(2, "0");
    };

    /**
     * Checks if the text is required to be truncated and broken into two lines based
     * on the available width of container. **Note: This is relevant to font size 18px or 1.125rem only **
     * @param text input text
     * @param widthOfElement element width in px
     */
    export const shouldTruncateToTwoLines = (
        text: string,
        widthOfElement: number
    ): boolean => {
        // This is arbitrary and based on tests to derive that every 9px increment ~= 1 char length.
        // But this applies to font size 18px or 1.125rem only
        const estimatedCharOnLine = Math.floor(widthOfElement / 9);
        return (
            text.length >= estimatedCharOnLine * 2 ||
            (text.split(" ").length === 1 && text.length > estimatedCharOnLine)
        );
    };

    /**
     * Performs a truncation by adding ellipsis in the middle of the text.
     * @param text input text
     * @param widthOfElement the dynamic width of the container in px
     * @param minDivSize the minimum container size (using size on mobileS viewport as baseline)
     * @param baselineCharLength the baseline amount of characters to be displayed before truncation kicks. This
     * will be increased if there is more space available (derived from `widthOfElement`)
     * @param incrementFactor the size (in px).
     */
    export const truncateOneLine = (
        text: string,
        widthOfElement: number,
        minDivSize: number,
        baselineCharLength: number,
        incrementFactor = 8 // Arbitrary based on tests and applies to font size of 18px or 1.125rem
    ): string => {
        let additionalCharAllowed = 0;
        if (widthOfElement > minDivSize) {
            additionalCharAllowed = Math.floor(
                (widthOfElement - minDivSize) / incrementFactor
            );
        }
        const allowedCharLength = baselineCharLength + additionalCharAllowed;
        if (allowedCharLength < text.length) {
            const thresholdIndex = Math.floor(allowedCharLength / 2); // We will cut it halfway to derive x num of characters
            return (
                text.substring(0, thresholdIndex) + // Use the first x
                " ... " +
                text.substring(text.length - thresholdIndex, text.length) // Use the last x
            );
        }

        return text;
    };

    /**
     * Returns the width of the text in the specified font
     * https://stackoverflow.com/questions/118241/calculate-text-width-with-javascript/21015393#21015393
     */
    export const getTextWidth = (text: string, font: string): number => {
        if (typeof document === "undefined") {
            return 0;
        }

        // create a reusable canvas object for better performance
        const canvas =
            (getTextWidth as any).canvas ||
            ((getTextWidth as any).canvas = document.createElement("canvas"));
        const context = canvas.getContext("2d");
        context.font = font;

        const metrics = context.measureText(text);
        return metrics.width;
    };

    export const maskValue = (value: string, options?: MaskAttributeProps) => {
        if (!value) {
            return value;
        }

        const {
            maskRange,
            unmaskRange,
            maskChar = "â€¢",
            maskRegex,
            maskTransformer,
        } = options;

        if (maskTransformer) {
            return maskTransformer(value);
        } else if (maskRegex) {
            return value.replace(maskRegex, maskChar);
        } else if (maskRange) {
            const { startIndex, endIndex } = determineStartAndEndIndex(
                maskRange[0],
                maskRange[1]
            );
            return (
                value.substring(0, startIndex) +
                maskChar.repeat(
                    value.substring(startIndex, endIndex + 1).length
                ) +
                value.substring(endIndex + 1)
            );
        } else if (unmaskRange) {
            const { startIndex, endIndex } = determineStartAndEndIndex(
                unmaskRange[0],
                unmaskRange[1]
            );
            return (
                maskChar.repeat(value.substring(0, startIndex).length) +
                value.substring(startIndex, endIndex + 1) +
                maskChar.repeat(value.substring(endIndex + 1).length)
            );
        }

        return value;
    };

    const determineStartAndEndIndex = (index0: number, index1: number) => {
        return index0 < index1
            ? { startIndex: index0, endIndex: index1 }
            : { startIndex: index1, endIndex: index0 };
    };
}
