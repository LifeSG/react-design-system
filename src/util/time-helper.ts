import dayjs from "dayjs";
import { StringHelper } from "./string-helper";

// =============================================================================
// INTERFACES
// =============================================================================
export type TimeFormat = "12hr" | "24hr";

export enum EPeriod {
    AM = "AM",
    PM = "PM",
}

export interface TimeValues {
    hour: string;
    minute: string;
    period: EPeriod;
}

// unexportable
interface TimeValuesPlain {
    hour: string;
    minute: string;
    period?: string;
}

// =============================================================================
// HELPER FUNCTIONS
// =============================================================================
export namespace TimeHelper {
    /**
     * Rounds time to the nearest interval, e.g 6:30 will be clamped to 6:00 when interval = 60
     * @param time the input time in HH:mm format
     * @param toNextHour to clamp to next interval instead, e.g. 6:30 will be clamped to 7:00 when interval = 60
     * @returns the time rounded to the nearest hour in HH:mm format
     */
    export const roundToNearestInterval = (
        time: string,
        interval: number, // Interval in minutes (e.g., 15 for 15 minutes, 60 for 1 hour)
        toNextInterval?: boolean
    ): string => {
        const [hoursStr, minutesStr] = time.split(":");
        const hours = parseInt(hoursStr, 10);
        const minutes = parseInt(minutesStr, 10);

        // Handle invalid inputs
        if (isNaN(hours) || isNaN(minutes) || minutes < 0 || minutes >= 60) {
            throw new Error("Invalid time format");
        }

        // Convert the time to total minutes
        const totalMinutes = hours * 60 + minutes;
        const remainder = totalMinutes % interval;

        // Calculate the rounded total minutes
        const roundedMinutes =
            remainder === 0
                ? totalMinutes
                : toNextInterval
                ? totalMinutes + (interval - remainder)
                : totalMinutes - remainder;

        // Convert the rounded total minutes back to hours and minutes
        const roundedHours = Math.floor(roundedMinutes / 60);
        const roundedMinutesWithinHour = roundedMinutes % 60;

        // Format the result as HH:mm, allowing for hours >= 24
        const formattedHours = roundedHours.toString().padStart(2, "0");
        const formattedMinutes = roundedMinutesWithinHour
            .toString()
            .padStart(2, "0");

        return `${formattedHours}:${formattedMinutes}`;
    };

    export const generateHourlyIntervals = (
        startTime: string,
        endTime: string,
        generatedFormat = "ha"
    ) => {
        const format = "HH:mm";
        let start = dayjs(startTime, format);
        let end = dayjs(endTime, format);

        if (start.isSame(end)) {
            end = end.add(1, "day");
        }

        const intervals: string[] = [];

        while (start.isBefore(end)) {
            intervals.push(start.format(generatedFormat));
            start = start.add(1, "hour");
        }

        return intervals;
    };

    export const getTimeValues = (
        format: TimeFormat,
        value: string | undefined
    ): TimeValues => {
        // Default value
        const timeValues: TimeValues = {
            hour: "",
            minute: "",
            period: EPeriod.AM,
        };

        if (!value) return timeValues;

        try {
            if (format === "24hr") {
                const plain = convertToPlain(value, format);
                // Add minute
                timeValues.minute = StringHelper.padValue(plain.minute);

                // Determine period, and convert to 12 hour value
                const hour = parseInt(plain.hour);

                if (Math.floor(hour / 12) === 0) {
                    timeValues.period = EPeriod.AM;
                    timeValues.hour =
                        hour === 0
                            ? "12"
                            : StringHelper.padValue(hour.toString());
                } else {
                    timeValues.period = EPeriod.PM;
                    timeValues.hour =
                        hour === 12
                            ? hour.toString()
                            : StringHelper.padValue((hour - 12).toString());
                }
            } else {
                const {
                    hour,
                    minute,
                    period = "",
                } = convertToPlain(value, format);

                timeValues.hour = StringHelper.padValue(hour);
                timeValues.minute = StringHelper.padValue(minute);
                timeValues.period =
                    period.toLowerCase() === "am" ? EPeriod.AM : EPeriod.PM;
            }

            return timeValues;
        } catch (error) {
            return timeValues;
        }
    };

    export const updateMinutes = (
        valueString: string,
        direction: "add" | "minus"
    ): string => {
        const currentValue = parseInt(valueString);

        if (isNaN(currentValue)) {
            return direction === "add" ? StringHelper.padValue("0") : "55";
        }

        /**
         * We would increment or decrement by steps of 5
         * Hence from 0 to 55, there are 12 steps
         */
        const steps = 12;

        /**
         * Look for last nearest step
         * E.g. value 4 would be step 0 = 0, value 6 would be step 1 = 5
         */
        const lastNearestStep = Math.floor(currentValue / 5);

        /**
         * Getting the next step value would have to incorporate a cycle of the upper
         * and lower limits. In this case upper = 12, lower = 0.
         * If one is to increment from step 12, it would then go back to 0.
         * If one is to decrement from step 0, it would then go back to 12.
         * Reference: https://dev.to/turneremma21/circular-access-of-array-in-javascript-j52
         */
        const nextStep =
            direction === "add"
                ? lastNearestStep + 1
                : currentValue % 5 === 0
                ? lastNearestStep - 1 // if it is a value on the step, decrement one more
                : lastNearestStep; // else the last nearest step would suffice
        const newStep = ((nextStep % steps) + steps) % steps;

        return StringHelper.padValue((newStep * 5).toString());
    };

    export const updateHours = (
        valueString: string,
        direction: "add" | "minus"
    ): string => {
        const currentValue = parseInt(valueString);

        if (isNaN(currentValue)) {
            return direction === "add" ? StringHelper.padValue("1") : "12";
        }

        const nextValue =
            direction === "add" ? currentValue + 1 : currentValue - 1;

        return nextValue <= 12 && nextValue > 0
            ? StringHelper.padValue(nextValue.toString())
            : nextValue === 13
            ? StringHelper.padValue("1")
            : "12";
    };

    export const convertTo24HourFormat = (values: TimeValues): string => {
        const hour = parseInt(values.hour);
        let hourString: string;

        if (values.period === EPeriod.PM) {
            hourString = hour === 12 ? hour.toString() : (hour + 12).toString();
        } else {
            hourString = hour === 12 ? "00" : values.hour;
        }

        return `${hourString}:${values.minute}`;
    };

    export const convertHourTo12HourFormat = (hourValue: string): string => {
        const hour = parseInt(hourValue);
        const hourString =
            hour % 12 === 0 ? (12).toString() : (hour % 12).toString();

        return StringHelper.padValue(hourString);
    };

    export const formatDisplayValue = (
        value: string | undefined,
        format: TimeFormat
    ): string => {
        try {
            const { hour, minute, period = "" } = convertToPlain(value, format);

            const paddedHour = StringHelper.padValue(hour);
            const paddedMinute = StringHelper.padValue(minute);

            let formatted = `${paddedHour}:${paddedMinute}`;

            if (format === "12hr") {
                formatted += period.toLowerCase();

                return formatted;
            }

            return formatted;
        } catch (error) {
            return "";
        }
    };

    export const toMinutesSeconds = (_seconds: number) => {
        const minutes = Math.floor(_seconds / 60);
        const seconds = _seconds % 60;

        return {
            minutes,
            seconds,
        };
    };

    // Converts h:mma/hh:mma to 24hr (eg. 13:00)
    export const to24Hour = (time: string) => {
        if (time?.includes("am") || time?.includes("pm")) {
            const [t, p] = time.split(/(am|pm)/i);
            const [hr, m] = t.split(":").map(Number);
            let h = hr;
            if (p === "pm" && h < 12) h += 12;
            if (p === "am" && h === 12) h = 0;
            return toTimeString(h, m);
        }
        return time; // No conversion if string alr has am/pm
    };

    // Generates an array of timings based on given startTime/interval/format
    export const generateTimings = (
        interval: number, // In minutes
        format: TimeFormat = "12hr",
        startTime?: string | undefined,
        endTime?: string | undefined // Inclusive
    ): string[] => {
        const timings = [];
        let currentMinutes = 0;
        let endMinutes = 1440 - interval; // Do not include next day's 12am

        // Convert startTime (h:mma) to minutes
        if (startTime) {
            currentMinutes = timeToMinutes(startTime);
        }
        if (endTime) {
            endMinutes = timeToMinutes(endTime);
        }

        while (currentMinutes <= endMinutes) {
            let hours = Math.floor(currentMinutes / 60);
            const minutes = currentMinutes % 60;

            if (format === "12hr") {
                const period = hours >= 12 ? "pm" : "am";

                hours = hours % 12;
                hours = hours ? hours : 12; // Convert hour 0 to 12

                const timeString = toTimeString(hours, minutes, period);
                timings.push(timeString);
            } else {
                const timeString = toTimeString(hours, minutes);
                timings.push(timeString);
            }

            currentMinutes += interval;
        }

        return timings;
    };

    // Return undefined = invalid field, "" = empty field, else returns h:mma
    export const parseInput = (
        input: string,
        format: TimeFormat = "12hr" // Returned format
    ): string | undefined => {
        if (input === "" || input === undefined) return input;

        const sanitizedInput = input.trim().toLowerCase();
        const timeRegex = /^(?:(\d{1,2})([:.])?(\d{2})?)?(a|p|am|pm)?$/;
        const match = timeRegex.exec(sanitizedInput);

        if (!match) return;

        let hours = parseInt(match[1] || "0", 10);
        const minutes = parseInt(match[3] || "0", 10);
        let period = match[4];

        // Ensure hours/mins are valid
        if (match[1] === undefined || hours > 24 || minutes > 59) return;

        // Convert single-character periods to full am/pm
        if (period === "a") {
            period = "am";
        } else if (period === "p") {
            period = "pm";
        }

        if (format === "24hr") {
            // Convert 12-hour input to 24-hour format if period is present
            if (period === "pm" && hours < 12) {
                hours += 12;
            } else if ((period === "am" && hours === 12) || hours === 24) {
                hours = 0; // Midnight case
            }

            // Return time in 24-hour format (HH:mm)
            return toTimeString(hours, minutes);
        }

        // Handle 24-hour times or AM/PM conversion
        if (period) {
            if (hours === 0 || hours === 24) {
                period = "am";
                hours = 12;
            } else if (hours > 12) {
                period = "pm";
                hours -= 12;
            }
        } else {
            period =
                // NOTE: 12 or 1-6 will default to pm for convenience
                hours === 0 || hours === 24 || (hours > 6 && hours < 12)
                    ? "am"
                    : "pm";
            hours = hours % 12 || 12;
        }

        // Format the time as h:mma
        const formattedTime = toTimeString(hours, minutes, period);

        return formattedTime;
    };

    export const findClosestFlooredTime = (
        inputTime: string | undefined,
        timeArray: string[] // Should already be sorted in ascending order
    ): string | undefined => {
        if (!inputTime) return inputTime;
        const flooredInputMinutes = timeToMinutes(inputTime);

        let closestTime = "";
        let minDifference = Infinity;

        for (const time of timeArray) {
            const timeInMinutes = timeToMinutes(time);
            const difference = timeInMinutes - flooredInputMinutes;

            // If the difference is negative or zero, update the closest time
            if (difference <= 0 && Math.abs(difference) < minDifference) {
                minDifference = Math.abs(difference);
                closestTime = time;
            }
            // If the difference becomes positive, we've passed the closest time
            else if (difference > 0) {
                break;
            }
        }

        return closestTime;
    };

    // Accepts both 12hr and 24hr formats
    export const timeToMinutes = (time: string) => {
        const [startHourMin, ampm] = time.toLowerCase().split(/(am|pm)/);
        const [startHoursRaw, startMinutes] = startHourMin
            .split(":")
            .map(Number);
        let startHours = startHoursRaw;

        if (ampm === "pm" && startHours !== 12) {
            startHours += 12; // Convert PM hours to 24-hour format
        }
        if (ampm === "am" && startHours === 12) {
            startHours = 0; // Convert 12am to 0 hours
        }

        return startHours * 60 + startMinutes;
    };
}

// =============================================================================
// NON-EXPORTABLES
// =============================================================================
const isValidHour = (hourString: string, format: TimeFormat): boolean => {
    const numValue = parseInt(hourString);
    return format === "24hr"
        ? numValue >= 0 && numValue <= 23
        : numValue >= 1 && numValue <= 12;
};

const isValidMinutes = (hourString: string): boolean => {
    const numValue = parseInt(hourString);
    return numValue >= 0 && numValue <= 59;
};

const isValidTimePeriod = (timePeriodString: string): boolean => {
    return (
        timePeriodString.toLowerCase() === "am" ||
        timePeriodString.toLowerCase() === "pm"
    );
};

const convertToPlain = (
    value: string | undefined,
    format: TimeFormat
): TimeValuesPlain => {
    const timeArr = value ? value.split(":") : [];
    const error = new Error("Invalid format");

    if (format === "12hr") {
        // Check format
        if (timeArr.length !== 2 || timeArr[1].length !== 4) throw error;

        const minute = timeArr[1].substring(0, 2);
        const period = timeArr[1].substring(2);

        // Validate hour, minute and time period values
        if (
            !isValidHour(timeArr[0], format) ||
            !isValidMinutes(minute) ||
            !isValidTimePeriod(period)
        ) {
            throw error;
        }

        return {
            hour: timeArr[0],
            minute,
            period: timeArr[1].substring(2),
        };
    } else {
        // Check format
        if (timeArr.length !== 2) throw error;

        // Validate hour and minute values
        if (!isValidHour(timeArr[0], format) || !isValidMinutes(timeArr[1])) {
            throw error;
        }

        return {
            hour: timeArr[0],
            minute: timeArr[1],
        };
    }
};

const toTimeString = (hours: number, minutes: number, period?: string) => {
    return period
        ? `${hours}:${minutes.toString().padStart(2, "0")}${period}`
        : `${hours.toString().padStart(2, "0")}:${minutes
              .toString()
              .padStart(2, "0")}`;
};
