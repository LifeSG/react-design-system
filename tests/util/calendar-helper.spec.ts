import dayjs from "dayjs";
import { generateCalendar } from "../../src/util/calendar-helper";

describe("GeneratCalendar", () => {
    it("should receive dayjs format value ", () => {
        const dayjsObject = dayjs();

        expect(generateCalendar(dayjsObject)).toBeTruthy();
    });
});
