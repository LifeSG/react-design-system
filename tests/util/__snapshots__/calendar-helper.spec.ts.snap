// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarHelper generateDays should receive dayjs format value  1`] = `
Array [
  Array [
    "2022-12-31T16:00:00.000Z",
    "2023-01-01T16:00:00.000Z",
    "2023-01-02T16:00:00.000Z",
    "2023-01-03T16:00:00.000Z",
    "2023-01-04T16:00:00.000Z",
    "2023-01-05T16:00:00.000Z",
    "2023-01-06T16:00:00.000Z",
  ],
  Array [
    "2023-01-07T16:00:00.000Z",
    "2023-01-08T16:00:00.000Z",
    "2023-01-09T16:00:00.000Z",
    "2023-01-10T16:00:00.000Z",
    "2023-01-11T16:00:00.000Z",
    "2023-01-12T16:00:00.000Z",
    "2023-01-13T16:00:00.000Z",
  ],
  Array [
    "2023-01-14T16:00:00.000Z",
    "2023-01-15T16:00:00.000Z",
    "2023-01-16T16:00:00.000Z",
    "2023-01-17T16:00:00.000Z",
    "2023-01-18T16:00:00.000Z",
    "2023-01-19T16:00:00.000Z",
    "2023-01-20T16:00:00.000Z",
  ],
  Array [
    "2023-01-21T16:00:00.000Z",
    "2023-01-22T16:00:00.000Z",
    "2023-01-23T16:00:00.000Z",
    "2023-01-24T16:00:00.000Z",
    "2023-01-25T16:00:00.000Z",
    "2023-01-26T16:00:00.000Z",
    "2023-01-27T16:00:00.000Z",
  ],
  Array [
    "2023-01-28T16:00:00.000Z",
    "2023-01-29T16:00:00.000Z",
    "2023-01-30T16:00:00.000Z",
    "2023-01-31T16:00:00.000Z",
    "2023-02-01T16:00:00.000Z",
    "2023-02-02T16:00:00.000Z",
    "2023-02-03T16:00:00.000Z",
  ],
  Array [
    "2023-02-04T16:00:00.000Z",
    "2023-02-05T16:00:00.000Z",
    "2023-02-06T16:00:00.000Z",
    "2023-02-07T16:00:00.000Z",
    "2023-02-08T16:00:00.000Z",
    "2023-02-09T16:00:00.000Z",
    "2023-02-10T16:00:00.000Z",
  ],
]
`;
